apisix/plugins/elasticsearch-logger.lua

```diff
@@ -0,0 +1,181 @@
+    local port = url_decoded.port
+    local index = conf.index .. "-" .. os.date('%Y.%m.%d')
+
+    core.log.info("sending a batch logs to ", conf.endpoint_addr .. index)
+
+    if not port then
+        if url_decoded.scheme == "https" then
+            port = 443
+        else
+            port = 80
+        end
+    end
+
+    local httpc = http.new()
+    httpc:set_timeout(conf.timeout * 1000)
+    local ok, err = httpc:connect(host, port)
+
+    if not ok then
+        return false, "failed to connect to host[" .. host .. "] port["
+            .. tostring(port) .. "] " .. err
+    end
+
+    if url_decoded.scheme == "https" then
+        ok, err = httpc:ssl_handshake(true, host, conf.ssl_verify)
+        if not ok then
+            return false, "failed to perform SSL with host[" .. host .. "] "
+                .. "port[" .. tostring(port) .. "] " .. err
+        end
+    end
+
+    local url_path = index .. "/_doc"
+    local user_pass = conf.user .. ":" .. conf.password
+    user_pass = "Basic " .. encode_base64(user_pass)
+    local httpc_res, httpc_err = httpc:request({
+        method = "POST",
+        path = url_path,
+        query = url_decoded.query,
+        body = log_message,
+        headers = {
+            ["Host"] = url_decoded.host,
+            ["Content-Type"] = "application/json",
+            ["Authorization"] = user_pass,
+        }
+    })
+
+    if not httpc_res then
+        return false, "error while sending data to [" .. host .. "] port["
+            .. tostring(port) .. "] " .. httpc_err
+    end
+
+    -- some error occurred in the server
+    if httpc_res.status >= 400 then
+        res =  false
+        err_msg = "server returned status code[" .. httpc_res.status .. "] host["
+            .. host .. "] port[" .. tostring(port) .. "] "
+            .. "body[" .. httpc_res:read_body() .. "]"
+    end
+
+    return res, err_msg
+end
+
+
+function _M.log(conf, ctx)
+    local metadata = plugin.plugin_metadata(plugin_name)
+    core.log.info("metadata: ", core.json.delay_encode(metadata))
+    local entry
+
+    if metadata and metadata.value.log_format
+       and core.table.nkeys(metadata.value.log_format) > 0
+    then
+        entry = log_util.get_custom_format_log(ctx, metadata.value.log_format)
+    else
+        entry = log_util.get_full_log(ngx, conf)
+    end
+
+    if batch_processor_manager:add_entry(conf, entry) then
+        return
+    end
+
+    -- Generate a function to be executed by the batch processor
+    local func = function(entries, batch_max_size)
+        local data, err
+
+        if batch_max_size == 1 then
+            data, err = core.json.encode(entries[1]) -- encode as single {}
+        else
+            local log_table = {}
+            for i = 1, #entries do
+                core.table.insert(log_table, core.json.encode(entries[i]))
+            end
+            data = core.table.concat(log_table, " ")  -- assemble multi items as string "{} {}"
```

Comment:
Just curious, are the space-separated JSONs going to be treated as separate events?

Else, just a suggestion - we can always use the elasticsearch bulk API (`endpoint/index/_bulk`). The indexing speed is much faster there.

Reply:
In fact, I would recommend just using the API for batch processing with ES. ES should be more read and less write.

Reply:
> Just curious, are the space-separated JSONs going to be treated as separate events?
> 
> Else, just a suggestion - we can always use the elasticsearch bulk API (`endpoint/index/_bulk`). The indexing speed is much faster there.

yes，using the bulk API is a good choice.I will try to improve it

Reply:
> In fact, I would recommend just using the API for batch processing with ES. ES should be more read and less write.

I don't agree with ES should be more read and less write.In simple scenarios, we don't need to worry about reading and writing. In a complex scene，we can use index lifecycle management to optimize. In fact, I think we should pay attention to its ecology，for ELKB.

Reply:
Consider the following reasons for multiple degrees of underwriting.
1. experience shows that ES write performance is not high and that high-frequency writes create performance problems.
2 For APISIX, writing data at high frequencies will also have performance problems, and APISIX provides batch process feature, and almost all logging plugins will follow this idea. I think combining the batch process of APISIX and the bulk of ES is a more suitable idea.